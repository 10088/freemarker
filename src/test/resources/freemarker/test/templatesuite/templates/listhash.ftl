<#setting booleanFormat='Y,N'>

<#macro listings maps>
  <#list maps as m>
    Map:
    
    [
    <#list m as k, v>
      ${k!'null'} = ${v!'null'}
    </#list>
    ]
  
    [
    <#list m as k, v>
      ${k!'null'} = ${v!'null'}<#sep>;</#sep> // @${k?index}=@${v?index}; ${k?itemParity}=${v?itemParity}; ${k?hasNext}=${v?hasNext}
    <#else>
      Empty
    </#list>
    ]
  
    {
    <#list m>
      [
      <#items as k, v>
        ${k!'null'} = ${v!'null'}<#sep>;</#sep> // @${k?index}=@${v?index}; ${k?itemParity}=${v?itemParity}; ${k?hasNext}=${v?hasNext}
      </#items>
      ]
    <#else>
      Empty
    </#list>
    }

  </#list>
</#macro>

Non-empty maps:

<@listings listables.hashEx2s />
<@listings [ listables.hashNonEx2 ] />

Empty maps:

<@listings listables.emptyHashes />

<#list { 'a': { 'aa': 11 }, 'b': { 'ba': 21, 'bb': 22 }, 'c': {} } as k1, v1>
  ${k1} @ ${k1?index}, ${v1?size}
  <#list v1 as k2, v2>
    ${k2} = ${v2}  @ ${k2?index} // inside ${k1} @ ${k1?index}, ${v1?size}
  </#list>
  ${k1} @ ${k1?index}, ${v1?size}
  --
</#list>
