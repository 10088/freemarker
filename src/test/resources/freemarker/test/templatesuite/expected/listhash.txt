
Non-empty maps:

    Map:
    
    [
      k1 = v1
      2 = v2
      k3 = v3
      null = v4
      Y = v5
      N = null
    ]
  
    [
      k1 = v1; // @0=@0; odd=odd; Y=Y
      2 = v2; // @1=@1; even=even; Y=Y
      k3 = v3; // @2=@2; odd=odd; Y=Y
      null = v4; // @3=@3; even=even; Y=Y
      Y = v5; // @4=@4; odd=odd; Y=Y
      N = null // @5=@5; even=even; N=N
    ]
  
    {
      [
        k1 = v1; // @0=@0; odd=odd; Y=Y
        2 = v2; // @1=@1; even=even; Y=Y
        k3 = v3; // @2=@2; odd=odd; Y=Y
        null = v4; // @3=@3; even=even; Y=Y
        Y = v5; // @4=@4; odd=odd; Y=Y
        N = null // @5=@5; even=even; N=N
      ]
    }

    Map:
    
    [
      k1 = v1
      2 = v2
      k3 = v3
      null = v4
      Y = v5
      N = null
    ]
  
    [
      k1 = v1; // @0=@0; odd=odd; Y=Y
      2 = v2; // @1=@1; even=even; Y=Y
      k3 = v3; // @2=@2; odd=odd; Y=Y
      null = v4; // @3=@3; even=even; Y=Y
      Y = v5; // @4=@4; odd=odd; Y=Y
      N = null // @5=@5; even=even; N=N
    ]
  
    {
      [
        k1 = v1; // @0=@0; odd=odd; Y=Y
        2 = v2; // @1=@1; even=even; Y=Y
        k3 = v3; // @2=@2; odd=odd; Y=Y
        null = v4; // @3=@3; even=even; Y=Y
        Y = v5; // @4=@4; odd=odd; Y=Y
        N = null // @5=@5; even=even; N=N
      ]
    }

    Map:
    
    [
      k1 = v1
      2 = v2
      k3 = v3
      null = v4
      Y = v5
      N = null
    ]
  
    [
      k1 = v1; // @0=@0; odd=odd; Y=Y
      2 = v2; // @1=@1; even=even; Y=Y
      k3 = v3; // @2=@2; odd=odd; Y=Y
      null = v4; // @3=@3; even=even; Y=Y
      Y = v5; // @4=@4; odd=odd; Y=Y
      N = null // @5=@5; even=even; N=N
    ]
  
    {
      [
        k1 = v1; // @0=@0; odd=odd; Y=Y
        2 = v2; // @1=@1; even=even; Y=Y
        k3 = v3; // @2=@2; odd=odd; Y=Y
        null = v4; // @3=@3; even=even; Y=Y
        Y = v5; // @4=@4; odd=odd; Y=Y
        N = null // @5=@5; even=even; N=N
      ]
    }

    Map:
    
    [
      k1 = 11
      k2 = 22
    ]
  
    [
      k1 = 11; // @0=@0; odd=odd; Y=Y
      k2 = 22 // @1=@1; even=even; N=N
    ]
  
    {
      [
        k1 = 11; // @0=@0; odd=odd; Y=Y
        k2 = 22 // @1=@1; even=even; N=N
      ]
    }


Empty maps:

    Map:
    
    [
    ]
  
    [
      Empty
    ]
  
    {
      Empty
    }

    Map:
    
    [
    ]
  
    [
      Empty
    ]
  
    {
      Empty
    }

    Map:
    
    [
    ]
  
    [
      Empty
    ]
  
    {
      Empty
    }

    Map:
    
    [
    ]
  
    [
      Empty
    ]
  
    {
      Empty
    }


  a @ 0, 1
    aa = 11  @ 0 // inside a @ 0, 1
  a @ 0, 1
  --
  b @ 1, 2
    ba = 21  @ 0 // inside b @ 1, 2
    bb = 22  @ 1 // inside b @ 1, 2
  b @ 1, 2
  --
  c @ 2, 0
  c @ 2, 0
  --
